### DISCUSSION TITLE
Auroraboot netboot caching for faster subsequent startup

### DISCUSSION THREAD

**Discussion by mevatron** [kairos-io/kairos#1827] [Q&A] [answered]
I am currently testing auroraboot for setting up an HA k3s cluster. Currently, I am running the following commands to get an Ubuntu 22.04 based cluster up and running:

Master node 1 (cluster init):

```
docker run --rm -ti -v $(pwd)/config-cluster-init.yaml:/config.yaml --net host quay.io/kairos/auroraboot --set "container_image=quay.io/kairos/kairos-ubuntu-22-lts:v2.4.0-k3sv1.27.3-k3s1" --cloud-config /config.yaml
```

Master node 2:

```
docker run --rm -ti -v $(pwd)/config-cluster-master-2.yaml:/config.yaml --net host quay.io/kairos/auroraboot --set "container_image=quay.io/kairos/kairos-ubuntu-22-lts:v2.4.0-k3sv1.27.3-k3s1" --cloud-config /config.yaml
```

Worker node 1:

```
docker run --rm -ti -v $(pwd)/config-cluster-worker-1.yaml:/config.yaml --net host quay.io/kairos/auroraboot --set "container_image=quay.io/kairos/kairos-ubuntu-22-lts:v2.4.0-k3sv1.27.3-k3s1" --cloud-config /config.yaml
```

Each of these step takes about 5 minutes to regenerate the ISO:

```
4:49PM INF Pulling container image 'quay.io/kairos/kairos-ubuntu-22-lts:v2.4.0-k3sv1.27.3-k3s1' to '/tmp/temp-rootfs' (local: false)
4:50PM INF Generating iso 'kairos' from '/tmp/temp-rootfs' to '/tmp/iso'
4:54PM INF Listening on :8080...
4:54PM INF Extracting netboot artifacts 'kairos' from '/tmp/iso/kairos.iso' to '/tmp/netboot'
4:54PM INF Start pixiecore
```

Is there a way to cache some of the steps, or a quicker way to switch a single session between different cloud configs?

Thanks, and great work on this cool project!
Will

**Comment by jimmykarily** [comment, marked as answer]
In case you haven't checked these options (eliminating the need for separate config file for each node):

- https://kairos.io/docs/examples/p2p_e2e/
- https://kairos.io/docs/examples/multi-node-p2p-ha/

To your question now, are you using auroraboot to netboot or just to generate the iso? If you are simply generating the iso then you can use the same iso to boot each node and use a different config at installation time. That's either using "manual install" or by passing the proper config when registering the node with the QR code, or whatever method you plan to use for installation.

If you are netbooting with auroraboot, I suspect you are going to have issues with all 3 instances running in the same network. That's because they advertise themselves as "netboot servers" over dhcp which will probably confuse the node on which one to use.

**Comment by mevatron** [comment]
@jimmykarily Thanks for the reply! I ended up going with the above as that 5 minute latency wasn't that bad considering I had to fiddle with the BIOS on each node before getting it to auroraboot :smile:, but I'll keep the p2p mode in mind for future decentralized clusters!
